In [2]:
# !pip install openpyxl

In [3]:
import pandas as pd

In [4]:
df = pd.read_excel(r'online_retail_combined.xlsx')

In [5]:
df.shape

(1048575, 8)

In [6]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   Invoice      1048575 non-null  object        
 1   StockCode    1048575 non-null  object        
 2   Description  1044203 non-null  object        
 3   Quantity     1048575 non-null  int64         
 4   InvoiceDate  1048575 non-null  datetime64[ns]
 5   Price        1048575 non-null  float64       
 6   Customer ID  811893 non-null   float64       
 7   Country      1048575 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 64.0+ MB


In [8]:
df.describe()

,Quantity,InvoiceDate,Price,Customer ID
count,1.048575e+06,1048575,1.048575e+06,811893.000000
mean,9.957525e+00,2010-12-27 19:44:22.911264768,4.627346e+00,15324.712265
min,-7.421500e+04,2009-12-01 07:45:00,-5.359436e+04,12346.000000
25%,1.000000e+00,2010-07-05 13:46:00,1.250000e+00,13971.000000
50%,3.000000e+00,2010-12-06 09:58:00,2.100000e+00,15260.000000
75%,1.000000e+01,2011-07-12 16:20:00,4.150000e+00,16795.000000
max,7.421500e+04,2011-12-04 13:15:00,3.897000e+04,18287.000000
std,1.335187e+02,NaN,1.228024e+02,1697.033034


# Checking Missing Values

In [9]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      4372
Quantity            0
InvoiceDate         0
Price               0
Customer ID    236682
Country             0
dtype: int64

In [10]:
(df.isnull().sum()/len(df))*100

Invoice         0.000000
StockCode       0.000000
Description     0.416947
Quantity        0.000000
InvoiceDate     0.000000
Price           0.000000
Customer ID    22.571776
Country         0.000000
dtype: float64

In [11]:
missing_report = pd.DataFrame({
    'Missing Count' : df.isnull().sum(),
    'Missing %' : (df.isnull().sum()/len(df))*100
})

In [12]:
missing_report.sort_values(by='Missing %',ascending=False)

,Missing Count,Missing %
Customer ID,236682,22.571776
Description,4372,0.416947
Invoice,0,0.000000
StockCode,0,0.000000
Quantity,0,0.000000
InvoiceDate,0,0.000000
Price,0,0.000000
Country,0,0.000000


# Identifying and removing the duplicates

In [13]:
df.duplicated().sum()

34150

In [14]:
df[df.duplicated()]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
383,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom
384,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
...,...,...,...,...,...,...,...,...
1048334,580469,23048,SET OF 10 LANTERNS FAIRY LIGHT STAR,1,2011-12-04 12:32:00,4.15,14583.0,United Kingdom
1048335,580469,21559,STRAWBERRY LUNCH BOX WITH CUTLERY,1,2011-12-04 12:32:00,2.55,14583.0,United Kingdom
1048349,580469,22079,RIBBON REEL HEARTS DESIGN,2,2011-12-04 12:32:00,1.65,14583.0,United Kingdom
1048510,580501,23438,RED SPOT GIFT BAG LARGE,2,2011-12-04 13:00:00,1.25,14546.0,United Kingdom


In [15]:
(df.duplicated().sum()/len(df))*100

3.2568008964547124

In [16]:
df = df.drop_duplicates()

In [17]:
# Found 34150 duplicated rows and removed them.

# Explore Cancellations

In [18]:
df[df['Invoice'].astype(str).str.startswith('C')].shape

(18872, 8)

In [19]:
cancel = df[df['Invoice'].astype(str).str.startswith('C')]
len(cancel)/len(df)*100

1.8603642457549843

In [20]:
df[df['Quantity']<0].head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia


In [21]:
#Around 1.86% transactions are cancellations.
#Cancellation invoices contain negative quantities

# Checking the DataTypes

In [22]:
df.dtypes

Invoice                object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
Price                 float64
Customer ID           float64
Country                object
dtype: object

In [23]:
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

# Revenue Column

In [24]:
df['Revenue'] = df['Quantity'] * df['Price']

In [25]:
df['Revenue'].describe()

count    1.014425e+06
mean     1.827768e+01
std      1.803112e+02
min     -7.718360e+04
25%      3.750000e+00
50%      9.950000e+00
75%      1.770000e+01
max      7.718360e+04
Name: Revenue, dtype: float64

# Missing value report

In [26]:
missing_report.style.background_gradient(cmap='Reds')

,Missing Count,Missing %
Invoice,0,0.000000
StockCode,0,0.000000
Description,4372,0.416947
Quantity,0,0.000000
InvoiceDate,0,0.000000
Price,0,0.000000
Customer ID,236682,22.571776
Country,0,0.000000


# Observations :

1) Dataset contains 1048575 records.
2) CustomerID had approx. 22.57% missing values.
3) Approx. 1.86% transactions are cancellations.
4) DupliDuplicate rows detected and removed.
5) Revenue column created for further analysis.

In [28]:
df_clean = df[~df['Invoice'].astype(str).str.startswith('C')]

In [29]:
df.to_csv("cleaned_retail.csv", index=False)